checks/frozen_deps.go

```diff
@@ -28,10 +31,76 @@ func init() {
 
 // FrozenDeps will check the repository if it contains frozen dependecies.
 func FrozenDeps(c *checker.CheckRequest) checker.CheckResult {
+	return checker.MultiCheckAnd(
+		lockFilesPresent,
+		workflowActionsPinned,
+	)(c)
+}
+
+// TODO: need to support Docker https://github.com/ossf/scorecard/issues/403
+
+// Check file content
+func onFileContent(path string, content []byte,
+	logf func(s string, f ...interface{})) bool {
+
+	// Structure for workflow config.
+	// We only retrieve what we need for logging.
+	// Github workflows format: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions
+	type YamlWorflowConfig struct {
+		Name string `yaml:name`
+		Jobs map[string]struct {
+			Name  string `yaml:name`
+			Steps []struct {
+				Name string `yaml:name`
+				Id   string `yaml:id`
+				Uses string `yaml:uses`
+			}
+		}
+	}
+
+	var workflow YamlWorflowConfig
+	err := yaml.Unmarshal(content, &workflow)
+	if err != nil {
+		panic(fmt.Errorf("!! frozen-deps - Cannot unmarshall file %v\n%v\n%v", path, content, string(content)))
+		return false
+	}
+
+	r := true
+	for jobName, job := range workflow.Jobs {
+		if len(job.Name) > 0 {
+			jobName = job.Name
+		}
+		for _, step := range job.Steps {
+			if len(step.Uses) > 0 {
+				// Ensure a hash at least as large as SHA1 is used (40 hex characters).
+				match, err := regexp.Match("^.*@[a-f\\d]{40,}$", []byte(step.Uses))
+				if err != nil {
+					panic(fmt.Errorf("!! frozen-deps - Regex failed for %v", path))
+				}
+				if !match {
+					r = false
+					logf("!! frozen-deps - %v has non-pinned dependency '%v' (job \"%v\")", path, step.Uses, jobName)
```

Comment:
We should consider updating the fn signature to return `(bool, error)`. That will not only avoid `panic` in this fn, but also, when we have any related details they can be returned so that they get added to `CheckResult.Details`. wdyt?

Reply:
yes right, this shouldbe added to details text. and not logged here.

Reply:
this logf function is the one that comes from the check_runner, which appends to the messages.